### PR TITLE
fix(snapshot): keep VM snapshot saves from being cancelled as device loss, and record orphaned captures for reclaim (#6961)

### LINKED ISSUE #6961: deviceSnapshot(capture, android): the liveness probe races the emulator console during `emu avd snapshot save`, cancelling a capture that succeeds and orphaning ~2 GB with no record
`adb emu avd snapshot save` freezes the emulator's console for ~3 s while it writes a multi-gigabyte VM snapshot. During that window the daemon's AVD-name console probe burns its 2000 ms budget, the device is declared disconnected, and `ExecutionTracker` **cancels the in-flight capture**. The tool returns `device_lost` / `confirmed-unavailable` — but the emulator finishes the save anyway, leaving a ~2 GB snapshot directory on disk with **no database row**: invisible to `restore`, and invisible to the archive accounting.

Intermittent and timing-dependent: 2 of 5 captures on the same emulator in one session.

## Repro (Android API 36, `am-api36-ga-arm64`, emulator-5556, server `0.0.71+ga0f51cc3e86c` = `main` @ `a0f51cc3e`)

```
getAndroid { deviceId: "emulator-5556" }
setToolEnabled { toolName: "deviceSnapshot", enabled: true }
deviceSnapshot { platform: "android", action: "capture" }

  -> isError: true
     {"code":"device_lost","deviceId":"emulator-5556",
      "sessionUuid":"2ae67df0-5aae-49f1-91c5-7edc55258f4f","reason":"confirmed-unavailable"}
```

The emulator never actually went away — `adb devices` lists `emulator-5556 device` throughout, and the `qemu-system-*` process is untouched.

## Daemon trace — the whole race in three seconds

```
19:15:03.730 [INFO] Capturing snapshot 'am-api36-ga-arm64_2026-09-13_19-15-03-727Z' for device emulator-5556
19:15:03.730 [INFO] Executing: adb -s emulator-5556 emu avd snapshot save am-api36-ga-arm64_2026-09-13_19-15-03-727Z
19:15:06.598 [DEBUG] AVD name resolution for emulator-5556 spent its 2000ms budget on the console probe;
                     skipping the property fallback
19:15:06.741 [INFO] [ExecutionTracker] Cancelled execution cfc52d8c-… for deviceSessionUuid=2ae67df0-…
                     (tool=deviceSnapshot, reason=device-disconnected:emulator-5556)
19:15:06.742 [ERROR] Failed to capture VM snapshot:
                     VM snapshot save failed for '…19-15-03-727Z': device-disconnected:emulator-5556
```

The liveness probe and the snapshot save contend for the **same emulator console**. The save holds it for longer than the probe's 2000 ms budget, so the probe's timeout is read as "device gone".

## The orphan it leaves

Despite the cancellation the emulator completed the write:

```
$ ls -d ~/.android/avd/am-api36-ga-arm64.avd/snapshots/*/ | grep -v default_boot
/Users/jason/.android/avd/am-api36-ga-arm64.avd/snapshots/am-api36-ga-arm64_2026-09-13_19-15-03-727Z/   # 2.0 GB

$ sqlite3 ~/.auto-mobile/auto-mobile.db \
    "select snapshot_name, size_bytes from device_snapshots where snapshot_name like 'am-api36%'"
(no rows)
```

So a failed capture silently costs 2.0 GB of disk that nothing will ever reclaim — the reclaim path in #6891 works from the record table, and there is no record.

## Notes for the fix

- A console probe timing out while *this daemon* is itself holding the console for a snapshot save is not evidence of disconnection. The liveness check should either be suppressed for the duration of a console-exclusive operation, or fall back to `adb devices` / the property path before concluding `confirmed-unavailable` (the log line says the property fallback was explicitly skipped).
- Either way the capture should not be cancelled by it, since the underlying `emu avd snapshot save` completes regardless.
- Independently, a capture that fails after the emulator wrote the payload should clean up the on-disk directory, or record it so the existing reclaim can collect it.

Related: #6960 (same tool, the surviving captures are then evicted by the 100 MB archive budget).

Filed by the nightly manual-test routine.

### PR DESCRIPTION
**Pull request by kaeawc**
Fixes a race where `adb emu avd snapshot save` holds the emulator console long enough to starve the daemon's AVD-name identity probe. `DevicePool` misread the resulting placeholder name as a different runtime taking the serial, quarantined it and cancelled the in-flight `deviceSnapshot` capture, even though the save succeeded on the device and left a ~2 GB payload with no record.

- Adds a shared `EmulatorConsoleBusyRegistry` (interface, in-memory implementation, fake). Console-exclusive VM snapshot save, load and delete mark the serial busy for their duration, and identity quarantine is skipped for a busy serial that `adb devices` still lists. Genuine absence detection is untouched and covered by a test.
- A capture that fails after the save command was dispatched records a `pendingReclaim` VM row, so the reclaim sweep from [#6891](https://github.com/kaeawc/auto-mobile/pull/6891) collects the orphaned in-AVD payload instead of it silently costing disk forever.

Complements [#6960](https://github.com/kaeawc/auto-mobile/issues/6960) (archive-budget eviction of the captures that do survive).

Closes [#6961](https://github.com/kaeawc/auto-mobile/issues/6961).

### CODE REVIEW

File: src/features/action/CaptureSnapshot.ts [hunk 1]
```diff
@@ -179,7 +187,10 @@ export class CaptureSnapshot implements SnapshotCaptureProvider {
 
       let result;
       try {
-        result = await this.adb.executeCommand(saveCommand, vmSnapshotTimeoutMs);
+        saveDispatched = true;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/features/action/CaptureSnapshot.ts]: Fixed in b2c251375: `saveDispatched` is set inside `AdbExecutor.execute(..., { beforeDispatch })`, after path resolution and immediately before the subprocess spawns; a pre-dispatch failure leaves VM_SNAPSHOT_SAVE_DISPATCHED unset (test added), so the prior row is preserved.

File: src/daemon/devicePool.ts [hunk 2]
```diff
@@ -6365,6 +6378,26 @@ export class DevicePool {
...
 
+  private shouldQuarantineUnresolvedEmulatorName(
+    pooled: PooledDevice,
+    discovered: Pick<BootedDevice, "deviceId" | "name" | "platform">,
+  ): boolean {
+    if (!this.hasUnresolvedEmulatorName(discovered)) {
+      return false;
+    }
+    if (!this.consoleBusyRegistry.isBusy(pooled.id)) {
+      return true;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/devicePool.ts]: Fixed in b2c251375: BootedDevice carries `consoleBusyDuringProbe`, stamped when the `emu avd name` probe fails, and quarantine gates on that carried flag rather than a live registry re-check. Tests: busy-at-probe with the registry idle by reconciliation → not quarantined; idle-at-probe → still quarantined.

File: src/server/deviceSnapshotManager.ts [hunk 3]
```diff
@@ -1391,46 +1463,66 @@ export async function captureDeviceSnapshot(
...
-          lastAccessedAt: timestamp,
-          sizeBytes,
-          manifest: captured.manifest,
-        });
-
-        return captured;
-      },
-    );
+      }
+      throw error;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/deviceSnapshotManager.ts:1590]: Fixed in b2c251375: recordFailedVmSnapshotReclaim reports whether it persisted, and notifySnapshotResources() runs before the rethrow when it did; a subscriber test asserts notifications/resources/updated fires on this path.

File: src/server/deviceSnapshotManager.ts [hunk 4]
```diff
@@ -347,6 +351,76 @@ async function resolveSnapshotSizeBytes(
...
+      `[DeviceSnapshot] Failed to measure orphaned VM snapshot '${snapshotName}' on AVD ` +
+        `'${device.name}': ${errorMessage(measureError)}`,
+      measureError,
+    );
+  }
+
+  const timestamp = now().toISOString();
+  try {
+    await snapshotRepository.insertSnapshot({
+      snapshotName,
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/deviceSnapshotManager.ts:437]: Fixed in cf0e88cb2: recordFailedVmSnapshotReclaim reads the existing row first; a row owned by a different deviceId is left untouched (warn) and the failed capture's own in-AVD payload is best-effort deleted. Test: AVD A's row unchanged after AVD B's dispatched failure.

File: src/server/deviceSnapshotManager.ts [hunk 5]
```diff
@@ -1391,46 +1465,69 @@ export async function captureDeviceSnapshot(
...
+
+          return captured;
+        },
+      );
+    } catch (error) {
+      if (
+        device.platform === "android" &&
+        device.deviceId.startsWith("emulator-") &&
+        wasVmSnapshotSaveDispatched(error)
+      ) {
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/deviceSnapshotManager.ts:1575]: Fixed in cf0e88cb2: captureDeviceSnapshot tracks that the VM capture itself succeeded and routes later persistence failures through the reclaim-recording path; if that insert also fails the payload is best-effort deleted. Two tests cover both cases.

File: src/utils/android-cmdline-tools/AndroidEmulatorClient.ts [hunk 6]
```diff
@@ -1477,6 +1496,7 @@ export class AndroidEmulatorClient implements AndroidEmulator {
       }
     } catch (error) {
       this.throwIfReadinessAborted(signal);
+      consoleBusyDuringProbe = this.consoleBusyRegistry.isBusy(deviceId);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/utils/android-cmdline-tools/AndroidEmulatorClient.ts]: Fixed in cf0e88cb2: the registry keeps a per-serial generation incremented on runExclusive entry and exit; the probe samples busy/generation before dispatch and flags consoleBusyDuringProbe if busy before, busy at failure, or the generation changed. A deferred-probe test lets the save fully complete before rejection and asserts the flag.

File: src/server/deviceSnapshotManager.ts [hunk 7]
```diff
@@ -347,6 +351,130 @@ async function resolveSnapshotSizeBytes(
...
+      `[DeviceSnapshot] Failed to measure orphaned VM snapshot '${snapshotName}' on AVD ` +
+        `'${device.name}': ${errorMessage(measureError)}`,
+      measureError,
+    );
+  }
+
+  const timestamp = now().toISOString();
+  try {
+    const existing = await snapshotRepository.getSnapshot(snapshotName);
+    if (existing && existing.deviceId !== device.deviceId) {
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/deviceSnapshotManager.ts]: Fixed in e19990f4f: the preserve check compares deviceName as well as deviceId, so a same-serial capture from a different AVD no longer replaces the original AVD's row (test added).

File: src/server/deviceSnapshotManager.ts [hunk 8]
```diff
@@ -347,6 +351,130 @@ async function resolveSnapshotSizeBytes(
...
+      deviceName: device.name,
+      platform: "android",
+      snapshotType: "vm",
+      includeAppData: true,
+      includeSettings,
+      createdAt: timestamp,
+      lastAccessedAt: timestamp,
+      sizeBytes,
+      pendingReclaim: true,
+      pendingReclaimReason: reason,
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/deviceSnapshotManager.ts:448]: Fixed in e19990f4f: restoreDeviceSnapshot throws an ActionableError for any pendingReclaim row before constructing a restore provider, and the archive listing reports `restorable: false` for them (tests for both).

File: src/utils/android-cmdline-tools/AndroidEmulatorClient.ts [hunk 9]
```diff
@@ -1477,6 +1498,10 @@ export class AndroidEmulatorClient implements AndroidEmulator {
       }
     } catch (error) {
       this.throwIfReadinessAborted(signal);
+      consoleBusyDuringProbe =
+        busyBeforeDispatch ||
+        this.consoleBusyRegistry.isBusy(deviceId) ||
+        generationBeforeDispatch !== this.consoleBusyRegistry.getGeneration(deviceId);
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/utils/android-cmdline-tools/AndroidEmulatorClient.ts]: Fixed in e19990f4f: the busy/generation overlap is computed once after the console probe, covering both the throw and empty-stdout cases; test: empty probe + empty property fallback + busy registry → consoleBusyDuringProbe true.

File: src/daemon/devicePool.ts [hunk 10]
```diff
@@ -6365,6 +6384,26 @@ export class DevicePool {
...
+  private shouldQuarantineUnresolvedEmulatorName(
+    pooled: PooledDevice,
+    discovered: Pick<BootedDevice, "deviceId" | "name" | "platform" | "consoleBusyDuringProbe">,
+  ): boolean {
+    if (!this.hasUnresolvedEmulatorName(discovered)) {
+      return false;
+    }
+    if (!discovered.consoleBusyDuringProbe) {
+      return true;
+    }
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/daemon/devicePool.ts:6402]: Fixed in e19990f4f: reconcilePooledIdentityResolution returns early on an unresolved observation, retaining the current quarantine state instead of lifting it; test quarantines via an idle probe then sends a busy unresolved probe and asserts the quarantine is retained.

File: src/server/deviceSnapshotManager.ts [hunk 11]
```diff
@@ -347,6 +352,133 @@ async function resolveSnapshotSizeBytes(
...
+      measureError,
+    );
+  }
+
+  const timestamp = now().toISOString();
+  try {
+    const existing = await snapshotRepository.getSnapshot(snapshotName);
+    if (
+      existing &&
+      (existing.deviceId !== device.deviceId || existing.deviceName !== device.name)
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/deviceSnapshotManager.ts:419]: Deferred to follow-up https://github.com/kaeawc/auto-mobile/issues/6970 (round 4 on the same mechanism; quoted verbatim there).

File: src/server/deviceSnapshotManager.ts [hunk 12]
```diff
@@ -347,6 +352,133 @@ async function resolveSnapshotSizeBytes(
...
+async function deleteUnrecordedVmSnapshot(
+  device: BootedDevice,
+  snapshotName: string,
+  avdSnapshots: AvdSnapshotOperations,
+  vmSnapshotTimeoutMs: number,
+  context: string,
+): Promise<void> {
+  try {
+    const outcome = await avdSnapshots.deleteVmSnapshot(
+      device.deviceId,
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/server/deviceSnapshotManager.ts:372]: Deferred to follow-up https://github.com/kaeawc/auto-mobile/issues/6970 (round 4 on the same mechanism; quoted verbatim there).

File: src/features/action/CaptureSnapshot.ts [hunk 13]
```diff
@@ -179,7 +187,14 @@ export class CaptureSnapshot implements SnapshotCaptureProvider {
 
       let result;
       try {
-        result = await this.adb.executeCommand(saveCommand, vmSnapshotTimeoutMs);
+        result = await this.consoleBusyRegistry.runExclusive(this.device.deviceId, () =>
+          this.adb.execute(saveCommand.split(" "), {
+            timeoutMs: vmSnapshotTimeoutMs,
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply src/features/action/CaptureSnapshot.ts:192]: Deferred to follow-up https://github.com/kaeawc/auto-mobile/issues/6970 (round 4 on the same mechanism; quoted verbatim there).
